src/TestResult.dfy

```diff
@@ -49,4 +51,93 @@ module TestResult {
     var resourceCount :- Externs.ParseNat(resourceCountStr);
     Success(TestResult(displayName, outcome, durationTicks, resourceCount))
   }
-}
\ No newline at end of file
+
+  method GroupTestResults(results: seq<TestResult>) returns (groupedResults: map<string, seq<TestResult>>) {
+    groupedResults := map[];
+    for resIndex := 0 to |results| {
+      var res := results[resIndex];
+      var name := res.displayName;
+      if name in groupedResults {
+        groupedResults := groupedResults[name := groupedResults[name] + [res]];
+      } else {
+        groupedResults := groupedResults[name := [res]];
+      }
+    }
+  }
+
+  function method TestResultStatistics(results: seq<TestResult>, f: TestResult -> int): Statistics.Statistics
+  {
+    match 0 < |results| {
+      case true =>
+        var min := Seq.Min(Seq.Map(f, results));
+        var max := Seq.Max(Seq.Map(f, results));
+        var mean := Statistics.Mean(Seq.Map(x => f(x) as real, results));
+        var stddev := Statistics.StdDev(Seq.Map(x => f(x) as real, results));
+        Statistics.Statistics(min as real, max as real, mean, stddev)
+      case false =>
+        Statistics.Statistics(0.0, 0.0, 0.0, 0.0)
+    }
+  }
+
+  function method TestResultDurationStatistics(results: seq<TestResult>): Statistics.Statistics
+  {
+    TestResultStatistics(results, (result: TestResult) => result.durationTicks as int)
+  }
+
+  function method TestResultResourceStatistics(results: seq<TestResult>): Statistics.Statistics
+  {
+    TestResultStatistics(results, (result: TestResult) => result.resourceCount)
+  }
+
+  method PrintTestResultStatistics(displayName: string, results: seq<TestResult>)
+  {
+    var timeStats := TestResultStatistics(results, (r: TestResult) => r.durationTicks as int);
+    var resStats := TestResultStatistics(results, (r: TestResult) => r.resourceCount as int);
+    print displayName, "\n";
+    print "  Time - ", Statistics.StatisticsToSeconds(timeStats).ToString(), "\n";
+    print "  Resource Count - ", resStats.ToString(), "\n";
+  }
+
+  method PrintAllTestResultStatistics(groupedResults: map<string, seq<TestResult>>)
+  {
+    var resultBatches := groupedResults.Items;
+    while resultBatches != {}
+      decreases |resultBatches|
+    {
+      var resultBatch :| resultBatch in resultBatches;
+      resultBatches := resultBatches - {resultBatch};
+      if 0 < |resultBatch.1| { // TODO: could prove that this never happens
+        PrintTestResultStatistics(resultBatch.0, resultBatch.1);
+      }
+    }
+  }
+
+  // TODO: there must be an easier way to map a function over a map
```

Comment:
A ma p would be this:

```
function method MapMap<K, V, V'>(m: map<K, V>, f: V -> V') : map<K, V'> {
  map k: K | k in m :: f(m[k])
}
```

But you want a fold, right?  I don't know of a good way to do that :'(

Reply:
Yeah, they're library functions. We really need traits on primitive and data types so that we can do arithmetic over sequences of arbitrary numeric things. :)

And, yes, that really is a fold! I'll update the comment.